File: SafeVote/fix_collections.js
const mongoose = require('mongoose');

const MONGO_URI = "mongodb+srv://stharunkumar069_db_user:siet%40123@cluster0.frcnaxx.mongodb.net/safevote?retryWrites=true&w=majority&appName=Cluster0";

async function fix() {
    try {
        await mongoose.connect(MONGO_URI);
        const db = mongoose.connection.db;

        const depts = ['CYBER', 'AIML'];
        const targetCollection = 'students';

        console.log("Starting consolidation...");

        for (let dept of depts) {
            const collection = db.collection(dept);
            const count = await collection.countDocuments();
            if (count > 0) {
                console.log(`Found ${count} documents in ${dept}. Moving to ${targetCollection}...`);
                const docs = await collection.find({}).toArray();

                for (let doc of docs) {
                    // Remove _id to avoid collision if necessary, or keep it if it's new
                    // Check if exists in target
                    const exists = await db.collection(targetCollection).findOne({ regNo: doc.regNo });
                    if (!exists) {
                        await db.collection(targetCollection).insertOne(doc);
                    } else {
                        // Update existing one with the one from dept collection
                        await db.collection(targetCollection).updateOne(
                            { regNo: doc.regNo },
                            { $set: { password: doc.password, name: doc.name, department: dept } }
                        );
                    }
                }
                console.log(`Finished moving ${dept}.`);
            }
        }

        console.log("Consolidation complete. The 'students' collection is now the source of truth.");
        process.exit(0);
    } catch (err) {
        console.error(err);
        process.exit(1);
    }
}

fix();

File: SafeVote/backend/elections/management/commands/fix_collections.py
from django.core.management.base import BaseCommand
from utils.database import get_db

class Command(BaseCommand):
    help = "Consolidate separate department collections into the students collection matching fix_collections.js"

    def handle(self, *args, **options):
        db = get_db()
        depts = ['CYBER', 'AIML']
        target_col = db['students']

        self.stdout.write("Starting department collection consolidation...")

        for dept in depts:
            if dept in db.list_collection_names():
                dept_col = db[dept]
                count = dept_col.count_documents({})
                if count > 0:
                    self.stdout.write(f"Found {count} documents in '{dept}'. Moving to 'students'...")
                    docs = list(dept_col.find({}))
                    for doc in docs:
                        reg_no = doc.get("regNo")
                        if reg_no is None:
                            continue
                        exists = target_col.find_one({"regNo": reg_no})
                        if not exists:
                            target_col.insert_one(doc)
                        else:
                            target_col.update_one(
                                {"regNo": reg_no},
                                {"$set": {
                                    "password": doc.get("password"),
                                    "name": doc.get("name"),
                                    "department": dept
                                }}
                            )
                    self.stdout.write(f"Finished moving '{dept}'.")

        self.stdout.write(self.style.SUCCESS("Consolidation complete. The 'students' collection is now the source of truth."))
